src/fparser/Fortran2003.py

```diff
@@ -552,16 +562,29 @@ class BracketBase(Base):
     <bracket-base> = <left-bracket-base> <something> <right-bracket>
     """
     def match(brackets, cls, string, require_cls=True):
-        i = len(brackets)/2
-        left = brackets[:i]
-        right = brackets[-i:]
+        bracket_len = len(brackets)/2
+        left = brackets[:bracket_len]
+        right = brackets[-bracket_len:]
         if string.startswith(left) and string.endswith(right):
-            line = string[i:-i].strip()
+            # We may have something like "(a + b)*(a - b)" so have
+            # to check - we start with one open bracket. If we reach
+            # zero open brackets before we get to the end then the
+            # opening bracket at the start of the string does not
+            # correspond to the closing bracket at the end of it.
+            num_open = 1
+            for idx in range(bracket_len, len(string)-bracket_len):
+                if string[idx:idx+bracket_len] == left:
+                    num_open += 1
+                elif string[idx:idx+bracket_len] == right:
+                    num_open -= 1
+                if num_open == 0:
+                    return
+            line = string[bracket_len:-bracket_len].strip()
             if not line:
                 if require_cls:
                     return
-                return left,None,right
-            return left,cls(line),right
+                return left, None, right
```

Comment:
024d782 I've re-structured `BracketBase.match()` so that it correctly rejects an empty expression "()". I think my Habakkuk changes had broken that because the check came after my new code that checked for closing braces within the expression. The modified line has only had spaces added and is still not touched. It will only be touched if `require_cls==False` and I don't know what that argument means. Therefore I propose not to try to get this line covered for the moment.